llext: add context save and restore

lyakh opened the pull request:
Some applications need to save LLEXT context, e.g. when suspending, to later restore it quickly without a full relinking. Add 2 functions for context saving and restoring.

lyakh committed:
SoC: Intel: ADSP: ACE30: add .imrdata to MMU definitions

On ACE30 platforms adding a section to the linker script isn't
enough, it should also be added to the xtensa_soc_mmu_ranges[] array.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

teburd commented on subsys/llext/llext.c:
```diff
@@ -324,3 +324,189 @@ void llext_bootstrap(struct llext *ext, llext_entry_fn_t entry_fn, void *user_da
 }
+
+/*
+ * Prepare a set of extension copies for future restoring. The user has copied
```
If all extensions are in an array, and we need to save/restore to different memory potentially (while maintaining the links) can we not use array indices instead of pointers for this internally? Wouldn't that simply this entire thing?

teburd commented on subsys/llext/llext.c:
```diff
@@ -324,3 +324,189 @@ void llext_bootstrap(struct llext *ext, llext_entry_fn_t entry_fn, void *user_da
+			for (k = 0; k < n_ext; k++) {
+				if (!strncmp(ext[k].name, ext[i].dependency[j]->name,
+				    sizeof(ext[k].name))) {
+					ext[i].dependency[j] = ext + k;
```
Right so rather than directly having a pointer here, can we not say that dependency[j] = k and ditch the pointers altogether?

lyakh replied:
@teburd extensions are usually created one by one. IIUC many use-cases will only have one extension, SOF typically uses several of them for a moderately complex stream, sometimes 2, sometimes 5, and we never destroy them. And there can be multiple streams in the system. So if you want to have them in an array, you'd have to have something like a pool of them, reallocating when it becomes too small... Not sure this would improve anything.

teburd commented on subsys/llext/llext.c:
```diff
@@ -324,3 +324,189 @@ void llext_bootstrap(struct llext *ext, llext_entry_fn_t entry_fn, void *user_da
+				goto free_locked;
+			}
+
+			next->dependency[j] = llext_by_name(next->dependency[j]->name);
```
This seems fairly costly, we're doing two loops and then walking the list of extensions comparing strings along the way. Had we used array indices instead here maybe we could have avoided all this work?

lyakh replied:
@teburd well, the loops are quite short. The most outer and the most inner loops are over all extensions, of which some use-cases have just 1, in SOF case we can have 5 or possibly up to 10. The middle loop over dependencies so far hardly executes more than once, not in SOF in any case, and I'm not sure if anybody else uses dependencies. So, it isn't that expensive. Not sure we need to switch to an array just because of this.

teburd reviewed (commented):
This seems like a lot of work to restore and save things, primarily it seems because the interdependence of extensions is all based on addresses which may change (if I'm understanding correctly). I wonder if instead we kept an array of pointers to extensions instead of a list pointer how this might change. I get that might be a larger change than wanted though. But at least then maybe we wouldn't need so many nested loops to do this work which.

The function descriptions could use with some elaboration on the intent, how they work, and how they are meant to be used.

pillo79 reviewed (commented):
Not a full review yet because I don't really understand what this PR is trying to solve... there are no tests, no documentation and a very short description of the PR intent which mirrors the commit. Can you please elaborate a little more?
For example, how is the new API supposed to be used? There's no `llext_save` operation to pair with `llext_restore`, so how would a user take advantage of this?

Also, I feel the Intel SOC changes should be in a separate PR, as I genuinely don't know anything about that :slightly_smiling_face:

lyakh commented:
@pillo79 at the top of this PR there's a note, that https://github.com/thesofproject/sof/pull/10028 refers to it, so that's a hint, that that is the SOF counterpart to it :-)
But no, I fully agree that this shouldn't just be "missing parts of some SOF work, that cannot be done in SOF, but have to be done in Zephyr." At the moment I'm afraid it's rather close to that. We can try to improve it.

To explain, here's the SOF use-case: SOF with Zephyr on Intel ADSP platforms run on those audio DSPs, under control of an operating system (Linux) on the host x86-64 CPU. During initialisation the host loads the firmware and LLEXT modules into shared DRAM, from where they are then copied into ADSP SRAM *on demand*. When no audio streaming is taking place the ADSP is shut down. But since DRAM is shared its content is preserved. Then when audio starts again, the ADSP is powered on. The firmware image is already in DRAM, so it doesn't have to be loaded from the OS filesystem, instead it's taken directly from DRAM. Then LLEXT objects have to be loaded, but the ADSP "doesn't know" that they're already in DRAM, or where they are there, or in which state (some of them have been linked while some others weren't). So until now we reload LLEXT completely, which is of course significant wasted time.

With these PRs I save all the information about loaded extensions in DRAM and on boot check if it's there, then I just restore it. Ideally I'd do all this in SOF, but some things aren't really possible there: specifically Zephyr LLEXT context should be allocated using `llext_alloc()`, which ATM is internal to Zephyr. Alternatively to this PR I could export that function and mostly do the rest internally. This also largely explains the asymmetry - why I don't need to save in Zephyr, but need to restore here: I can allocate DRAM buffers in SOF, while Zephyr LLEXT doesn't know about them, so it's easy for SOF to copy the entire Zephyr LLEXT context to DRAM. That's also why I need `struct llext_elf_sect_map`. I only preferred not to dive into dependencies, because they're pretty internal to Zephyr. But SOF cannot really restore Zephyr LLEXT context - at least because we cannot call `llext_alloc()`.

pillo79 commented:
Thanks for the explanation @lyakh, I had in fact not noticed the link yesterday! For reference, the "save" code I mentioned is currently [in SOF's `llext_manager_store_to_dram`](https://github.com/thesofproject/sof/blob/a511a6115970d7bc0010e61acfa843acc4158907/src/library_manager/llext_manager_dram.c#L34).

That is really neat, and would be great to have it in Zephyr. However, it relies on SOF storing each loader along with its extension, which is not mandated by Zephyr.

Not sure how to handle this. Options I see ahead:
* We can implement a proper save/restore in Zephyr for `struct llext` only, but... that's not enough for SOF.
* To implement saving loaders as well, we need to link extensions with their loaders somehow; however, we then rely on the user not releasing them. Maybe a new "I promise they are static" Kconfig?
* The option I would definitely not consider is... the situation in the current PR. That `store_to_dram` function in SOF _needs_ to have access to any field in the structs, making this setup extremely brittle to any Zephyr change (and we've been down that road!). Also, how should any other user take advantage of this? How should we reliably test it?

@lyakh @teburd Thoughts?

lyakh commented:
> Some thoughts…
> 
> It would be better to align a broader zephyr usage with the sof usage of llext to get the benefits of it being open source.
> 
> So imagine instead of dram it is flash that images are being saved to and restored from. I think that scenario opens up the flow for a lot more people. Getting that in zephyr, with docs and tests helps avoid bit rot and increase collaboration.
> 
> Serializing pointers is always problematic. So if that is what we are doing as I believe it is then yeah using an array and indexes solves this.
> 
> Testing could be as simple as saving and restoring extensions to another addressable area as I believe the dram is addressable unless I’m mistaken @lyakh?
> 
> If hooks are needed before/after the save and restore that too could be done here and have tests and such.

@teburd @pillo79 sure, I'd be happy to make this useful for many - but unfortunately I'm not sure how to make this useful for other hypothetical cases. If we had at least 2 actual real life users, we could make it usable by both, but we don't have a second user and we don't even know if we ever get one... But let's try to see how to improve this:
We need to save and restore multiple things:
- LLEXT objects, including
- exported symbol tables
- loaders, potentially including
- section maps
- dependencies
- user per-LLEXT context

At run-time LLEXT objects are linked in a list and I think it's better than an array - elements can be allocated and freed individually at random locations in the list.
Each LLEXT has an array of pointers to dependencies. This is ok as an array because we don't expect to add elements to it or remove them from it dynamically - once all the dependencies are found, they should stay there.
LLEXT and loader objects aren't linked by design.
LLEXT objects are managed by LLEXT, loader objects are managed by the user.
We could make a Zephyr LLEXT function to save all LLEXT-managed data to user-provided buffers - everything from the above list except loaders and user context. But the user will have to calculate the sizes. LLEXT also doesn't know which loader corresponds to which LLEXT. So, the user would provide an array of pointers to loaders, an array of pointers to extensions - in the same order, so there's always a 1:1 correspondence, and buffers for all of them... That would be pretty cumbersome.

lyakh committed:
llext: add context save and restore

Some applications need to save LLEXT context, e.g. when suspending,
to later restore it quickly without a full relinking. Add 2 functions
for context saving and restoring. Since these functions are likely to
change in the future, put them in llext_experimental.c, which depends
on CONFIG_LLEXT_EXPERIMENTAL and is disabled by default.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

pillo79 reviewed (commented):
Thanks for the Discord discussion. A few minor remarks from me, but no blockers.

lyakh commented on include/zephyr/llext/loader.h:
```diff
@@ -29,7 +29,10 @@ extern "C" {
 
 /** @cond ignore */
-struct llext_elf_sect_map; /* defined in llext_priv.h */
+struct llext_elf_sect_map {
+	enum llext_mem mem_idx;
+	size_t offset;
+};
```
...alternatively we could export a function, returning `sizeof(struct llext_elf_sect_map)` which would be ugly too, but at least wouldn't expose the structure itself

pillo79 replied:
Yes, sorry for the mixup - I meant `llext_internal.h` under `include/zephyr/llext/` :man_facepalming:

lyakh committed:
llext: rename _llext_list to llext_list

Beginning symbols or structure field names with an underscore isn't a
widely used style in Zephyr, remove the leading underscore from
_llext_list.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

teburd reviewed (approved):
My more significant concerns have been addressed. It's likely a bit of a refactoring will be needed in the future to better address the XIP and DRAM cases that are similar here.

For XIP and MMU-less architectures in particular more work is needed, for the DRAM and MMU case we'd need to supply the linked addresses to the platform so the MMU can be programmed to map those addresses appropriately I imagine.

In any case, there's no reason to -1 this as its useful today.

pillo79 reviewed (approved):
Agree, LGTM as well - docs included. Thanks!
